src/transformers/cache_utils.py

```diff
@@ -447,6 +447,123 @@ def batch_select_indices(self, indices: torch.Tensor):
+        # device because performing torch.index_select on the CPU is very slow
+        del self.beam_idx
+        self.beam_idx = beam_idx.clone()
+
+    def update(
+        self,
+        key_states: torch.Tensor,
+        value_states: torch.Tensor,
+        layer_idx: int,
+        cache_kwargs: Optional[Dict[str, Any]] = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        Updates the cache with the new `key_states` and `value_states` for the layer `layer_idx`.
+        Parameters:
+            key_states (`torch.Tensor`):
+                The new key states to cache.
+            value_states (`torch.Tensor`):
+                The new value states to cache.
+            layer_idx (`int`):
+                The index of the layer to cache the states for.
+            cache_kwargs (`Dict[str, Any]`, `optional`):
+                Additional arguments for the cache subclass. No additional arguments are used in `OffloadedCache`.
+        Return:
+            A tuple containing the updated key and value states.
+        """
+        # Update the number of seen tokens
+        if layer_idx == 0:
+            self._seen_tokens += key_states.shape[-2]
+
+        # Update the cache
+        if len(self.key_cache) <= layer_idx:
+            self.key_cache.append(key_states)
+            self.value_cache.append(value_states)
+            self.original_device.append(key_states.device)
+            self.evict_previous_layer(layer_idx)
+        else:
+            key_tensor, value_tensor = self[layer_idx]
+            self.key_cache[layer_idx] = torch.cat([key_tensor, key_states], dim=-2)
+            self.value_cache[layer_idx] = torch.cat([value_tensor, value_states], dim=-2)
+
+        return self.key_cache[layer_idx], self.value_cache[layer_idx]
+
+    @classmethod
+    def from_legacy_cache(cls, past_key_values: Optional[Tuple[Tuple[torch.FloatTensor]]] = None) -> "OffloadedCache":
+        """Converts a cache in the legacy cache format into an equivalent `DynamicCache`."""
+        cache = cls()
+        if past_key_values is not None:
+            for layer_idx in range(len(past_key_values)):
+                key_states, value_states = past_key_values[layer_idx]
+                cache.key_cache.append(key_states)
+                cache.value_cache.append(value_states)
+                cache.original_device.append(key_states.device)
+        return cache
+
+    def to_legacy_cache(self) -> Tuple[Tuple[torch.Tensor], Tuple[torch.Tensor]]:
+        """Converts the `OffloadedCache` instance into the its equivalent in the legacy cache format."""
+        legacy_cache_lst = []
+        for layer_idx in range(len(self)):
+            original_device = self.original_device[layer_idx]
+            self.key_cache[layer_idx] = self.key_cache[layer_idx].to(original_device)
+            legacy_cache_lst.append((self.key_cache[layer_idx], self.value_cache[layer_idx]))
+        return tuple(legacy_cache_lst)
```

Comment:
I would overload these two functions with a `NotImplementedError` exception -- they are exclusively used for backwards compatibility, and we would like to get rid of them in the future 🤗

Reply:
I have removed these implementations but did it the way it is recommended in the Python docs and left a comment in the code with a pointer to the docs.